Add Sidecar log level parameter to GS specification

aLekSer opened the pull request:
Add ability to specify `logrus` log level in yaml configuration and apply this log level on GameServer creation step.

Closes  #971 .

aLekSer commented:
What is left: documentation update and UT.

aLekSer commented:
Build failed because I should have performed `make gen-api-docs`
```
Step #15 - "tests": includes/website.mk:96: recipe for target 'test-gen-api-docs' failed
```

markmandel commented on pkg/apis/agones/v1/gameserver.go:
```diff
@@ -135,6 +135,8 @@ type GameServerSpec struct {
 	// Scheduling strategy. Defaults to "Packed".
 	Scheduling apis.SchedulingStrategy `json:"scheduling,omitempty"`
+	// LogLevel specifies the log level for gameserver sidecar.
+	LogLevel string `json:"loglevel,omitempty"`
```
Random thought: Is "loglevel" too generic? Considering it's just for the sdk sidecar - should we pick a name that's more specific? `sdk-server-loglevel` (seems long), not sure what the right answer is here,  figured I would ask the question first.

Just wondering if the name might be confusing.

aLekSer replied:
yes, you are right, that would suit better.

aLekSer commented:
Flaky E2E test:
```
Step #20: --- FAIL: TestFleetScaleUpEditAndScaleDown (0.01s)
Step #20: --- FAIL: TestFleetScaleUpEditAndScaleDown/Use_fleet_Patch_false (156.24s)
Step #20: fleet_test.go:120: 
Step #20: Error Trace:	fleet_test.go:120
Step #20: Error: Expected nil, but got: &errors.errorString{s:"timed out waiting for the condition"}
Step #20: Test: TestFleetScaleUpEditAndScaleDown/Use_fleet_Patch_false
```

markmandel commented on site/content/en/docs/Reference/gameserver.md:
```diff
@@ -49,6 +49,8 @@ spec:
     # Minimum consecutive failures for the health probe to be considered failed after having succeeded.
     # Defaults to 3. Minimum value is 1
     failureThreshold: 3
+  # SdkServer Log Level configuration. Could be any of panic, fatal, error, warn, warning, info, debug and trace. Defaults to info
```
we should `feature` wrap these items.

aLekSer replied:
Done

markmandel commented on install/yaml/install.yaml:
```diff
@@ -860,6 +882,17 @@ spec:
                             type: integer
                             minimum: 1
                             maximum: 65535
+                    loglevel:
```
Pretty sure this validation also needs to be changed, with the rename?

aLekSer replied:
yes, that's true

roberthbailey commented on install/helm/agones/templates/crds/_gameserverspecvalidation.yaml:
```diff
@@ -92,6 +92,17 @@ properties:
               maximum: 65535
+      sdk-server-loglevel:
+        type: string
+        enum:
```
I don't think we should expose so many log levels. This seems to be externalizing the fact that we are using logrus and would make it nearly impossible to change our logging library in the future. I feel like there are really three categories that people want:

1. Default (what we have today, which in basically the info level)
2. Errors and above
3. Debugging - everything

What if we abstracted the log levels from logrus and just offered something smaller like those three? Alternatively, we could give people a way to specifically disable the ping logs since those are the ones that cause the most noise in the log file.

aLekSer replied:
Yes, less level would be more concise. I will do the change cause that would actually give more space for the changes in the future.

aLekSer commented:
/assign @aLekSer

markmandel commented:
Looks like the CRD docs need to be manually updated, before it will pass tests.

markmandel commented:
Ah, might actually be:
```
go: git.apache.org/thrift.git@v0.0.0-20180902110319-2566ecd5d999: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /go/pkg/mod/cache/vcs/83dba939f95a790e497d565fc4418400145a1a514f955fa052f662d56e920c3e: exit status 128:
	fatal: unable to access 'https://git.apache.org/thrift.git/': Failed to connect to git.apache.org port 443: Connection timed out
```

roberthbailey commented on install/helm/agones/templates/crds/_gameserverspecvalidation.yaml:
```diff
@@ -92,6 +92,17 @@ properties:
+        type: string 
+        description: |
+                  sdkServerLogLevel has three options:
+                  - "Info" (default) output SDK server messages of all log levels except debug messages
```
Suggestions for making this a bit more consistent:

```
gameserverSidecarLogLevel has three options:
- "Info" (default) The game server sidecar will output all messages except for debug messages
- "Debug" The game server sidecar will output all messages including debug messages
- "Error" The game server sidecar will only output error messages

roberthbailey commented on install/helm/agones/templates/crds/_gameserverspecvalidation.yaml:
```diff
@@ -92,6 +92,17 @@ properties:
               type: integer
               minimum: 1
               maximum: 65535
+      sdkServerLogLevel:
```
You use sdk server in some places and game server sidecar in others. I'm not sure which is better, but we should be consistent one way or the other. I'm a bit partial to game server sidecar, since users of agones won't realize that the sidecar has to do with the SDK and because when you are looking at the logs in a k8s cluster those are the words used to name the container we inject into the pod with the game server container. 

I'd suggest making this field be `gameserverSidecarLogLevel`.

markmandel replied:
So this took me back - I didn't realise that was what we called the sdk server sidecar container!
`"agones-gameserver-sidecar"` - in retrospect, possibly not the best name! But agree, consistency is key.

I guess my only question is - do we ever thing we'd add another sidecar of some sort? In which case, should we be specific about which sidecar's log level we are manipulating?

roberthbailey replied:
> I guess my only question is - do we ever thing we'd add another sidecar of some sort? 

Are you suggesting making it `sidecarLogLevel` for now (shorter) or are you saying that we might have another sidecar for gameservers and that `gameserverSidecarLogLevel` isn't specific enough?

roberthbailey replied:
Maybe the gameserver prefix is unnecessary?

markmandel replied:
My concern was that "sidecar" could be too generic - that's why I was thinking sdk-server.

My other concern would be that "sidecar" is a well known K8s pattern, but I'm not sure if its knowledge we can assume across all our users?

roberthbailey replied:
I'd suggest either `agonesGameserverSidecarLogLevel` (if we want to exactly match the name of the sidecar container) or `gameserverSidecarLogLevel` (if we want a shorter field name). I'm inclined to use the longer name if we aren't going to shorten it all the way to "sidecar".

markmandel replied:
I think `agones` is redundant, since we're on an Agones GameServer. Ditto for `gameserverSidecarLogLevel` - we are in a gameServer already.

Maybe I wasn't clear - I prefer `sdkServerLogLevel` because it's specific to which thing it is being set to. "sidecar" could essentially be anything, and is also confusing to those who don't know Kubernetes patterns. What if the user creates their own sidecars, are we somehow overwriting their log level? If we ever add another sidecar for Agones GameServers (not likely, but possible), now we can't configure log levels for each one individually, because we don't have a name for each one.

Does that make more sense?

aLekSer replied:
I also like sdkServerLogLevel more.
However I can change parameter to gameserverSidecarLogLevel if it would be more natural to the users.

markmandel replied:
I had an idea this morning in the shower :shower: 

I think we should have a logging section. Looking at commentary in #1036 , it's possible we will have more than 1 sidecar. In which case, we likely don't want multiple top spec level logging config params.

So howabout this:

```yaml
spec:
  # ... other fields ... 
  logging:
    sdkServer: "warning"
```

Then if we have to add more logging parameters, they aren't peppered through the config, they can be scoped under this `logging` element.

WDYT?

markmandel replied:
Or maybe that should be `loggingLevel` ? (or `level` should be a sub-element to `logging` ?) Not 100% sure. But I think this is on a better path.

aLekSer commented:
Updated this PR as per latest Mark comment, updated the docs also.
If we need to change the ENV variables like the following it would be another PR.
```
        - name: SIDECAR_IMAGE # overwrite the GameServer sidecar image that is used
```

markmandel commented on examples/gameserver.yaml:
```diff
@@ -35,7 +35,7 @@ spec:
+    # portPolicy has three options:
     # - "Dynamic" (default) the system allocates a free hostPort for the gameserver, for game clients to connect to
     # - "Static", user defines the hostPort that the game client will connect to. Then onus is on the user to ensure that the
     # - "Passthrough" dynamically sets the `containerPort` to the same value as the dynamically selected hostPort.
```
Does this need an example in it, like we have in the fleet.yaml?

aLekSer replied:
Added new parameter to `example/gameserver.yaml`

markmandel commented on examples/gameserver.yaml:
```diff
@@ -35,7 +35,7 @@ spec:
+    # portPolicy has three options:
     # - "Dynamic" (default) the system allocates a free hostPort for the gameserver, for game clients to connect to
     # - "Static", user defines the hostPort that the game client will connect to. Then onus is on the user to ensure that the
     # - "Passthrough" dynamically sets the `containerPort` to the same value as the dynamically selected hostPort.
```
Should we also have a log level example in this file?

aLekSer replied:
Yes, definitely. I think we can omit updating fleet.yaml? Or we should update both?

aLekSer commented:
Updated PR and tested with different gameserver and fleet configurations.

roberthbailey commented on install/helm/agones/templates/crds/_gameserverspecvalidation.yaml:
```diff
@@ -92,6 +92,21 @@ properties:
               maximum: 65535
+      logging:
+        type: object
+        title: Define logging level for game server sidecars
```
nit: Maybe say "logging levels for agones system components"

roberthbailey commented on pkg/apis/agones/v1/gameserver.go:
```diff
@@ -133,12 +133,20 @@ type GameServerSpec struct {
 	Template corev1.PodTemplateSpec `json:"template"`
 }
 
+// Logging contains log levels for Gameserver Pod containers
```
nit: Suggest replacing "Gameserver Pod containers" with "Agones system containers"

roberthbailey commented on pkg/sdkserver/sdkserver.go:
```diff
@@ -192,8 +193,21 @@ func (s *SDKServer) Run(stop <-chan struct{}) error {
+	if err == nil {
+		s.logger.Logger.SetLevel(level)
+	} else {
+		s.logger.WithError(err).Info("Specified wrong Logging.SdkServer")
```
what log level is actually being used by logrus in this case? I assume it's info, in which case we should also set s.logLevel to "info" so that they are consistent.

roberthbailey replied:
on second thought, is there a reason that we store loglevel in the struct? why not make it a local variable?

aLekSer replied:
yes, you are right it should be simplified

aLekSer replied:
doing it as a member of struct was intended to support changing the logLevel on the go, by configuring GS Spec.

aLekSer commented:
Applied these comments. @roberthbailey you can review one more time.

aLekSer commented:
I have merged master to the branch.

aLekSer commented:
Updated publishVersion to new one:
{{% feature publishVersion="1.1.0" %}}

Also I have found an issue with `make gen-api-docs` that if PR moves through the release, then it doesn't see new changes and doesn't update versions accordingly. 
So it was needed to perform an extra step before the command:
`git checkout master ../site/content/en/docs/Reference/agones_crd_api_reference.html`

aLekSer commented:
Build failed for some strange reason - no status comment was generated here on `github` PR:
```
Status	
 Build failed
Status detail	
failed unmarshalling build config cloudbuild.yaml: proto: google.devtools.cloudbuild.v1.Build.timeout|"timeout": google.protobuf.Duration: invalid duration value "1h"
Build id	
0867e0f5-6a9b-48b8-92bb-f9abb0d96c16
Image	
—
Trigger	
Push to github_pr_1007 branch (agones)
```

roberthbailey commented on pkg/apis/agones/v1/gameserver.go:
```diff
@@ -133,12 +133,20 @@ type GameServerSpec struct {
-	// Scheduling strategy. Defaults to "Packed".
+	// Scheduling strategy. Defaults to "Packed"
 	Scheduling apis.SchedulingStrategy `json:"scheduling,omitempty"`
+	// Logging specifies the log level for gameserver sidecars
```
Can you make this match the comment below, e.g. "Logging specifies the log levels for Agones system containers"?

aLekSer replied:
Will update in a moment to:
```
 Logging specifies log levels for Agones system containers
```

roberthbailey commented on pkg/apis/agones/v1/gameserver.go:
```diff
@@ -133,12 +133,20 @@ type GameServerSpec struct {
 
+// Logging contains log levels for Agones system containers
+type Logging struct {
+	// SdkServer is a log level for SDK server (sidecar) logs
```
Please add `Defaults to "Info"` at the end of this comment.

roberthbailey commented on pkg/apis/agones/v1/gameserver_test.go:
```diff
@@ -152,12 +155,14 @@ func TestGameServerApplyDefaults(t *testing.T) {
 					InitialDelaySeconds: 11,
 					PeriodSeconds:       12,
 				},
+				sdkServerLogLevel: "Info",
```
this test is "defaults are already set" so the log level should be set in the gameserver as well as in the expected result.

aLekSer replied:
Done

roberthbailey commented on pkg/apis/agones/v1/gameserver_test.go:
```diff
@@ -205,17 +212,19 @@ func TestGameServerApplyDefaults(t *testing.T) {
 					},
-					Health: Health{Disabled: true},
+					Health:  Health{Disabled: true},
+					Logging: Logging{SdkServer: "Debug"},
```
rather than piggyback on an existing test case ("convert from legacy single port to multiple" to test applying a custom log level, please add a new test case that specifically tests setting the log level.

aLekSer replied:
OK, adding a separate test

roberthbailey commented on site/content/en/docs/Reference/fleet.md:
```diff
@@ -57,6 +57,11 @@ spec:
       health:
         initialDelaySeconds: 30
         periodSeconds: 60
+{{% feature publishVersion="1.0.0" %}}
```
This should now be 1.1.0

aLekSer commented:
Applied comments and added `agones_crd_api_reference.html`
